feat: [SimpleWebTransport] Add ability to specify a custom client port (for reverse proxy scenarios not using port 80/443)

aallbrig opened the pull request:
Hello Mirror devs -- very excited to be able to make a contribution to Mirror's source..!

Here is the summary of changes
- Adds new "User Specified Port" option to the SimpleWebTransport component
- The client websocket request uses this new port only if "Client Use Default Port" and the new port has a value.

Context:
1. I want to reverse proxy through a locally installed NGINX
1. I installed `nginx` locally on my macbook using homebrew (`brew`)
1. homebrew NGINX doesn't make it easy to bind to privileged ports like port 80 or 443 (it's cool because I can have NGINX bind to ports `8080` and/or `8443` instead).

The privileged port issue is a problem because the current way Mirror supports reverse proxies is via the `ClientUseDefaultPort` boolean flag and that flag ultimately only supports sending a `ws://` or `wss://` connection to port 80 or port 443, respectively.

This PR adds functionality so that the Mirror user can alternatively specify a custom port to connect the client websocket connection to. This will support my use case where I have NGINX listening on port 8443.

You can see this addition in action
<img width="526" alt="Screenshot 2023-11-10 at 10 19 28 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/0a18ee25-24e4-4fc7-a8b0-d9a8d1965172">

<img width="1512" alt="Screenshot 2023-11-10 at 10 26 35 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/bca109cd-5271-42c2-be12-1911dfac10d2">

More context: 
- https://discord.com/channels/343440455738064897/1171946245586157628/1171946245586157628

MrGadget1024 reviewed (commented):
Leave this open so I remember, but I'm going to handle this in a different way.

aallbrig commented:
@MrGadget1024 @miwarnec As per [a discussion in the websockets channel in the Mirror discord](https://discord.com/channels/343440455738064897/516693507151495168/1172583136148865195), here is a (rough) draft of what a custom editor might look like. If the mirror user wants to custom a client port for simple web transport (SWT) then they check a checkbox. Upon doing so will present a dropdown. If they select "specify port" then an input field will be made available to enter a custom port.

<img width="496" alt="Screenshot 2023-11-15 at 10 31 35 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/3e5a3737-2d22-4f0b-b060-beaaf4dcb969">
<img width="502" alt="Screenshot 2023-11-15 at 10 31 40 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/8d0468c3-1f35-48be-aa44-e4c7505e83ad">
<img width="497" alt="Screenshot 2023-11-15 at 10 31 44 AM" src="https://github.com/MirrorNetworking/Mi
<img width="502" alt="Screenshot 2023-11-15 at 10 31 53 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/80594f06-8cde-44cd-9ccd-1780e5c76c84">
rror/assets/3106250/4e5b593a-efd4-4c94-9929-76dec59f106f">

James-Frowen commented:
Something like this would be better 
![image](https://github.com/MirrorNetworking/Mirror/assets/23101891/3a6b5a17-12fe-4b74-8d78-039f41cd9491)
![image](https://github.com/MirrorNetworking/Mirror/assets/23101891/fa654129-1284-4746-a67a-f7c4ba3022af)
![image](https://github.com/MirrorNetworking/Mirror/assets/23101891/63d4b153-57be-4e28-ba9f-de0b269ddd79)
*with client use ssl*
![image](https://github.com/MirrorNetworking/Mirror/assets/23101891/eda6bb4a-9f6e-4036-8be6-51b8ba46f1bb)

```cs

namespace Mirror.SimpleWeb
{
    [Serializable]
    public struct ClientPortSettings
    {
        public ClientPortOptions Options;
        public ushort CustomPort;
    }

    public enum ClientPortOptions
    {
        DefaultPort,
        SameAsServer,
        CustomPort,
    }
}
namespace Mirror.SimpleWeb.EditorScripts
{
#if UNITY_EDITOR

    [CustomPropertyDrawer(typeof(ClientPortSettings))]
    public class ClientPortSettingsDrawer : PropertyDrawer
    {
        public override float GetPropertyHeight(SerializedProperty property, GUIContent label)
        {
            // force exapnd
            property.isExpanded = true;

            var optionHeight = EditorGUI.GetPropertyHeight(property.FindPropertyRelative(nameof(ClientPortSettings.Options)));
            var portHeight = EditorGUI.GetPropertyHeight(property.FindPropertyRelative(nameof(ClientPortSettings.CustomPort)));
            var spacing = EditorGUIUtility.standardVerticalSpacing;
            return optionHeight + spacing + portHeight;
        }

        public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
        {
            var optionProp = property.FindPropertyRelative(nameof(ClientPortSettings.Options));
            var portProp = property.FindPropertyRelative(nameof(ClientPortSettings.CustomPort));

            var optionHeight = EditorGUI.GetPropertyHeight(optionProp);
            var portHeight = EditorGUI.GetPropertyHeight(portProp);
            var spacing = EditorGUIUtility.standardVerticalSpacing;

            position.height = optionHeight;
            EditorGUI.PropertyField(position, optionProp);
            position.y += spacing + optionHeight;
            position.height = portHeight;

            var option = (ClientPortOptions)optionProp.enumValueIndex;
            if (option == ClientPortOptions.DefaultPort || option == ClientPortOptions.SameAsServer)
            {
                var port = 0;
                if (property.serializedObject.targetObject is SimpleWebTransport swt)
                {
                    if (option == ClientPortOptions.DefaultPort)
                        port = swt.clientUseWss ? 443 : 80;
                    else
                        port = swt.port;
                }

                var wasEnabled = GUI.enabled;
                GUI.enabled = false;
                EditorGUI.IntField(position, new GUIContent("Client Port"), port);
                GUI.enabled = wasEnabled;
            }
            else
            {
                EditorGUI.PropertyField(position, portProp);
            }
        }
    }
#endif
}

```

aallbrig commented:
@James-Frowen The solution feels so much better. I locally tested the "default port," "server port," and "custom port" (both http and https) scenarios locally. 

<img width="1512" alt="Screenshot 2023-11-15 at 4 44 05 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/8cb6a039-a890-440e-82a4-ced740906549">
<img width="1512" alt="Screenshot 2023-11-15 at 4 44 12 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/aa6ddb4d-fdc4-4339-b34d-d46041cc376c">

PS I have this repo that contains instructions on how to set up a simulation of SSL termination using self signed certs and NGINX (https://github.com/aallbrig/experiment-localhost-ssl-termination-with-mirror). The repo is a WIP but wanted to share what I'm using to test out scenarios.

aallbrig commented:
@James-Frowen I extended the code to allow the client to specify a custom path as well.

<img width="496" alt="Screenshot 2023-11-15 at 6 12 44 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/56e81a9f-9f49-412d-80ca-eaa87594587d">
<img width="1512" alt="Screenshot 2023-11-15 at 6 20 51 PM" src="https://github.com/MirrorNetworking/Mirror/a
<img width="1512" alt="Screenshot 2023-11-15 at 6 20 59 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/060e2852-2d04-4898-bdcb-9b9e00afc0f5">
ssets/3106250/7e22a8ee-9b04-4921-83ba-48276d15e8e6">

I tested using this NGINX directive:
```
        location /test_websocket_path {
            rewrite ^/test_websocket_path(/.*)?$ / break;

            proxy_pass http://localhost:7778; # Proxy pass to Mirror Unity server (expected to be ran in Unity Editor if localhost)

            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_set_header Host $host;
            proxy_read_timeout 86400;
        }

```
The directive rewrite the `custom path` to the one the `SimpleWebTransport` listens to by default (aka `/`). This allows Mirror users to specify which `port` and `path` the client attempts to listen to. This NGINX directive reverse proxies both the port and path to the SWT server (port `7778` path `/`). You can see in the attached screenshots a successful reverse proxy to my Mirror game server running in the Unity Editor.

aallbrig commented:
<img width="606" alt="Screenshot 2023-11-17 at 10 17 01 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/3f0a06a6-862f-4db6-8e27-55072182a77e">
<img width="615" alt="Screenshot 2023-11-17 at 10 17 13 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/16a42705-f6ac-48a3-a96c-438088356fc7">
<img width="612" alt="Screenshot 2023-11-17 at 10 17 32 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/51d714a4-66ce-4cc4-bc56-f9eb80ca7152">
<img width="613" alt="Screenshot 2023-11-17 at 10 17 41 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/2ebed4d1-a599-4f64-bbe0-9c4d452327aa">
<img width="599" alt="Screenshot 2023-11-17 at 10 20 33 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/00d424ee-cd26-4971-94af-c41953f9d103">
<img width="611" alt="Screenshot 2023-11-17 at 10 20 39 AM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/868ef130-059a-4e0a-8fc8-62f129c36407">

Changed the wording and added tool tips for the options (couldn't get screenshots of the tooltips)

MrGadget1024 reviewed (changes requested):
Please make an Editor folder under Transports/SimpleWeb and move the ClientWebsocketSettingsDrawer to that folder and revert changes to Mirror.Editor.asmdef

![image](https://github.com/MirrorNetworking/Mirror/assets/9826063/77b7c12e-5ccf-431a-a11e-5b0f2087bb30)

MrGadget1024 commented:
I'm not understanding what this is supposed to do:

![image](https://github.com/MirrorNetworking/Mirror/assets/9826063/bae18821-32eb-4f17-92a0-8980119acbef)

aallbrig commented:
@MrGadget1024 Thanks for reviewing this code. I moved the editor changes into a newly created simplewebtransport.editor folder.

> I'm not understanding what this is supposed to do...

Essentially the `ClientWebsocketSettings.PathOption` controls whether to optionally create a websocket URI that contains a custom path. This is useful for reverse proxy scenarios where the user wants to use the default `/` path for something else other than connecting to a mirror game server; instead of having the websocket URI connect to path `/`, the mirror user can specify another path to connect to.

Consider the following nginx snippet:

```nginx
server {
    server_name localhost;
    listen 8443 ssl;
    ssl_certificate      /opt/homebrew/etc/nginx/ssl/server.crt;
    ssl_certificate_key  /opt/homebrew/etc/nginx/ssl/server.key;

    access_log  /opt/homebrew/etc/nginx/logs/ssl_termination_site.access.log;
    sendfile        on;

    # Web Game Client served at /
    root   /opt/homebrew/etc/nginx/html/GameClient;
    index  index.html;

    # Game Server Connection Reverse Proxy
    location /connect {
        rewrite ^/connect(/.*)?$ / break;
        proxy_pass http://localhost:7778; # Proxy pass to Unity server (expected to be ran in Unity Editor)

        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_read_timeout 86400;
    }
}
```

This nginx snippet allows it to serve a game client at `https://localhost:8443` (aka `https://localhost:8443/`) and reverse proxy `https://localhost:8443/connect` to `http://localhost:7778/` (which in my case is a Mirror game server running in Unity Editor). Without the ability to customize the client path for the websocket connection then the websocket connection would always go to `/` in the reverse proxy.

aallbrig commented:
<img width="480" alt="Screenshot 2023-11-19 at 3 53 59 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/58beef0e-0989-402f-99b7-7edb9fd2c652">
<img width="752" alt="Screenshot 2023-11-19 at 3 55 23 PM" src="https://github.com/MirrorNetworking/Mirror/assets/3106250/25de11e1-5f3f-4ff6-bb37-b39e8ab0d155">

Here is the path option in action used with the nginx snippet I shared above.

MrGadget1024 commented:
You put the new Editor folder too deep.  I asked for it to be at Transports/SimpleWeb.  Please move it up and revert any changes to AsmDef's.  I did so here, and it works without touching any AsmDef's.

MrGadget1024 commented:
As for custom client path, I think that's really a corner case.  The more common case is to have the website on 443 and the reverse proxy on some other port at the same domain.  I'd like you to remove that part, perhaps to a separate PR.

James-Frowen commented:
If someone is using default port 443, they are likely to  be also using a path so that the reverse proxy can also host other things.

The majority of use case would **either** use the port, or the path. but there is no good reason to use both at once.

aallbrig commented:
@MrGadget1024 I removed the custom path code and reverted the edits to other asmdef I had done in previous code iterations (sorry about that). Let me know what other changes you would like to see to the contribution.

> If someone is using default port 443, they are likely to be also using a path so that the reverse proxy can also host other things.
>
> The majority of use case would either use the port, or the path. but there is no good reason to use both at once.

I find this a little confusing. I don't exactly understand how using a custom path or a custom port would imply anything about the other port or path, respectively? Apologizes for the ignorance. I trust the guidance though.

MrGadget1024 reviewed (changes requested):
Please remove Transports/SimpleWeb/Editor/SimpleWebTransport.Editor.asmdef as it's not needed.

MrGadget1024 commented on Assets/Mirror/Transports/SimpleWeb/Editor/ClientWebsocketSettingsDrawer.cs:
```diff
@@ -0,0 +1,68 @@
+
+        private void DrawPortSettings(Rect position, SerializedProperty property)
+        {
+            var portOptionProp = property.FindPropertyRelative(websocketPortOptionName);
+            var portProp = property.FindPropertyRelative(customPortName);
+            var portOptionHeight = EditorGUI.GetPropertyHeight(portOptionProp);
+            var portHeight = EditorGUI.GetPropertyHeight(portProp);
+            var spacing = EditorGUIUtility.standardVerticalSpacing;
+            var wasEnabled = GUI.enabled;
+
+            position.height = portOptionHeight;
+
+            EditorGUI.PropertyField(position, portOptionProp, portOptionLabel);
+            position.y += spacing + portOptionHeight;
+            position.height = portHeight;
+
+            var portOption = (WebsocketPortOption)portOptionProp.enumValueIndex;
```
Replace `var` with explicit types, here and elsewhere.

aallbrig commented:
@MrGadget1024 Done done and done :)

aallbrig commented:
Synced this fork/branch with the latest changes from MirrorNetworking/Mirror.

aallbrig commented:
Added the commit suggestions. The code should be ready for re-review :)

InfinityDevTech commented:
Glad this is being implemented, this matches my situation exactly lol

aallbrig commented:
@InfinityDevTech Good to hear! I hope this will be useful to others.

aallbrig commented:
Synced this fork/branch with the latest changes from MirrorNetworking/Mirror.
